# Day 7 Reading Journal

This journal includes several required exercises, but it is meant to encourage active reading more generally.  You should use the journal to take detailed notes, catalog questions, and explore the content from Think Python deeply.

Reading: Think Python Chapter 11, 12

**Due: Monday, February 13 at 12 noon**



## [Chapter 11](http://www.greenteapress.com/thinkpython/html/thinkpython012.html)


**Quick check:** In about one sentence using your own words, what is a dictionary?

A dictionary is a data strucutre that maps a unique key to each value(s) where the key is immutable. 

### Exercise 11.2  

Dictionaries have a method called [`get`](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict) that takes a key and a default value. If the key appears in the dictionary, `get` returns the corresponding value; otherwise it returns the default value. For example:

```
>>> h = histogram('a')
>>> print h
{'a': 1}
>>> h.get('a', 0)
1
>>> h.get('b', 0)
0
```

Use `get` to write `histogram` more concisely. You should be able to eliminate the `if` statement. Add unit tests for your histogram implementation.

In [32]:
import pprint
import doctest

def histogram(s):
    """
    Precondition : string s
    Postcondition : Returns a dictionary of characters and frequencies.
    
    >>> histogram('aaaaaaaa')
    {'a': 8}
    
    >>> {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1} == histogram('abcdef')
    True
    
    >>> pprint.pprint(histogram('abcdef'))  #pprint sorts the dictionary for us
    {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1}
    """
    d = dict()
    for c in s:
        d[c] = d.get(c,0)+1
    return d


doctest.run_docstring_examples(histogram, globals(),verbose=True)

Finding tests in NoName
Trying:
    histogram('aaaaaaaa')
Expecting:
    {'a': 8}
ok
Trying:
    {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1} == histogram('abcdef')
Expecting:
    True
ok
Trying:
    pprint.pprint(histogram('abcdef'))  #pprint sorts the dictionary for us
Expecting:
    {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1}
ok


### Exercise 11.4  

Modify `reverse_lookup` so that it builds and returns a list of all keys that map to `v`, or an empty list if there are none. Add unit tests for your implementation.

In [51]:
import doctest

def reverse_lookup(d, v):
    """
    Precondition: dictionary d, and value v
    Postcondition: Returns a list of all keys that maps to v
    
    >>> reverse_lookup({'a': 1, 'b': 1, 'c': 1}, 1)
    ['a', 'b', 'c']
    
    >>> reverse_lookup({'a': 1, 'b': 1, 'c': 1}, 2)
    []
    
    """
    all_keys = [] 
    for k in d:
        if d[k] == v:
            all_keys.append(k)
    all_keys.sort()
    return all_keys
    raise ValueError
    
doctest.run_docstring_examples(reverse_lookup, globals(),verbose=True)

Finding tests in NoName
Trying:
    reverse_lookup({'a': 1, 'b': 1, 'c': 1}, 1)
Expecting:
    ['a', 'b', 'c']
ok
Trying:
    reverse_lookup({'a': 1, 'b': 1, 'c': 1}, 2)
Expecting:
    []
ok


If you'd like to learn more about errors and exceptions, you can check out the [Python tutorial](https://docs.python.org/3/tutorial/errors.html) or read ahead to [Appendix A](http://www.greenteapress.com/thinkpython2/html/thinkpython2021.html) of Think Python. If you choose to use doctest for your unit testing, it can also [deal with exceptions](https://docs.python.org/3/library/doctest.html#what-about-exceptions).

**Quick check** What type of objects can be used as keys to a dictionary, i.e. what property must they have?

only immutable objects like integers and strings can be used(i.e. they have to be hashble). Lists can't be used because they mutable.

### Exercise 11.6 (modified)

Create a memoized version of your Levenshtein distance function from Day 7. What kind of performance change do you see?

Optional: If you'd like to get some quantitative results, you could check out the [timeit](https://docs.python.org/3/library/timeit.html) module

Note: You can also study Fibonacci here if you prefer.

In [95]:
# my function from day 7 (no memoization)
from timeit import default_timer as timer

def lavenshtein_distance(str1, str2):
    global count
    count += 1  # counts total number of recursive calls
    
    cost = 0

    if(len(str1) == 0):
        return len(str2)
    if(len(str2) == 0):
        return len(str1)

    if(str1[-1] == str2[-1]):
        cost = 0
    else:
        cost = 1

    # minimum of delete from str1, str2, or from both.
    return min(lavenshtein_distance(str1[:-1], str2)+1,
               lavenshtein_distance(str1, str2[:-1])+1,
               lavenshtein_distance(str1[:-1], str2[:-1])+cost)

count = 0  # using it as a global variable
start = timer()
print('lavenshtein distnace : ', lavenshtein_distance('abcdefga', 'gfedcbaa'))
end = timer()
print('Time Elapsed : ', end - start, 'sec')
print('Total Recursive Calls : ', count)

lavenshtein distnace :  6
Time Elapsed :  0.21827613599998585 sec
Total Recursive Calls :  398593


In [119]:
# added memoization

from timeit import default_timer as timer

def lavenshtein_distance(str1, str2):
    global count
    count += 1  # counts total number of recursive calls
    cost = 0
    
    if(len(str1) == 0):
        memo[(str1,str2)] = len(str2) #indexing using tuples
        return len(str2)
    if(len(str2) == 0):
        memo[(str1,str2)] = len(str1)
        return len(str1)

    if(str1[-1] == str2[-1]):
        cost = 0
    else:
        cost = 1

    if((str1,str2) in memo):
        return memo[(str1,str2)]
    else:
        # minimum of delete from str1, str2, or from both.
        memo[(str1,str2)] = min(lavenshtein_distance(str1[:-1], str2)+1,
                   lavenshtein_distance(str1, str2[:-1])+1,
                   lavenshtein_distance(str1[:-1], str2[:-1])+cost)
        return memo[(str1,str2)]

count = 0  # using it as a global variable
memo = {}
start = timer()
print('lavenshtein distnace : ', lavenshtein_distance('abcdefga', 'gfedcbaa'))
end = timer()
print('Time Elapsed : ', end - start, 'sec')
print('Total Recursive Calls : ', count)

lavenshtein distnace :  6
Time Elapsed :  0.0006069760001992108 sec
Total Recursive Calls :  193


We observe that the number of recursive calls has siginficantly decreased from 398593 to 193 for the same task! And the time elpased has also significantly decreased from 0.22 sec to 0.0006 sec. This is a siginficant increase in the program efficiency!

## [Chapter 12](http://www.greenteapress.com/thinkpython2/html/thinkpython2013.html)

**Quick check:** In about one sentence using your own words, what is a tuple?

It's a immutable sequence of things

### Chapter 12.4  

Many of the built-in functions use variable-length argument tuples. For example, `max` and `min` can take any number of arguments:

```
>>> max(1,2,3)
3
```

But `sum` does not.

```
>>> sum(1,2,3)
TypeError: sum expected at most 2 arguments, got 3
```

Write a function called ```sumall``` that takes any number of arguments and returns their sum. 

Write unit tests for your function. Do I actually need to keep saying this? Let's assume it's always a good idea :)

In [12]:
import doctest

def sumall(*args):
    """
    Precondition : any number of arguments a,b,c,...
    Postcondition : Return sum of arguments a,b,c,...
    
    >>> sumall(1,2,3)
    6
    
    >>> sumall(1,2,3,4,5)
    15
    
    >>> sumall(0,0,0,0)
    0
    
    """
    sum = 0
    for value in args:
        sum += value
    return sum
    
    
doctest.run_docstring_examples(sumall, globals(),verbose=True)

Finding tests in NoName
Trying:
    sumall(1,2,3)
Expecting:
    6
ok
Trying:
    sumall(1,2,3,4,5)
Expecting:
    15
ok
Trying:
    sumall(0,0,0,0)
Expecting:
    0
ok


If you're interested in more flexible ways to pass arguments to functions, check out the [Python tutorial](https://docs.python.org/3/tutorial/controlflow.html#more-on-defining-functions). For instance, you can also use keyword arguments, which are collected into a dictionary just like `*` gathers variable numbers of positional arguments into a tuple.

This pattern is very common for defining functions with complex optional behaviors in Python, and you will often see definitions like:

```
def my_func(required_argument1, *arguments, **keywords):
    ...
```

### Exercise

Write a function `sort_by_last_letter` that takes a list of words and returns a new list with the words sorted alphabetically by the _last letter_ in the word. Hint: use the **Decorate, Sort, Undecorate** pattern. Write unit tests for your function.

In [31]:
import doctest


def sort_by_last_letter(words):
    """
    Precondition : list of words
    Postcondition : a new list of words sorted alphabetically by the last letter
    
    >>> sort_by_last_letter(['abcd','abc','abbb','a'])
    ['a', 'abbb', 'abc', 'abcd']
    
    >>> sort_by_last_letter(['aaad','aaac','aacd','ab'])
    ['ab', 'aaac', 'aaad', 'aacd']
    """
    new_words = []
    for word in words:
        new_words.append(word[::-1])
    new_words.sort()
    for i, word in enumerate(new_words):
        new_words[i] = word[::-1]
    return new_words

doctest.run_docstring_examples(sort_by_last_letter, globals(),verbose=True)

Finding tests in NoName
Trying:
    sort_by_last_letter(['abcd','abc','abbb','a'])
Expecting:
    ['a', 'abbb', 'abc', 'abcd']
ok
Trying:
    sort_by_last_letter(['aaad','aaac','aacd','ab'])
Expecting:
    ['ab', 'aaac', 'aaad', 'aacd']
ok


**Quick check** Give an example of when you might use each sequence type:

- tuple

- list

- string

1) Tuple : I would use it when I want to pass multiple variables to a function or return many numbers from a function(using gather and scatter). Also it can be used as keys for dictionaries.

2) List : I would use it when for bascially any sequences except for cases when I must use tuple (dictionary). 

3) String : I would use it for a sequence of characters that doesn't necessarily need to be mutable.

### Exercise 12.1 

Write a function called `most_frequent` that takes a string and prints the letters in decreasing order of frequency. Find text samples from several different languages and see how letter frequency varies between languages. Compare your results with the tables at http://en.wikipedia.org/wiki/Letter_frequencies. 

Allen's solution (try it on your own first): http://greenteapress.com/thinkpython2/code/most_frequent.py. 

In [56]:
import doctest

def most_frequent(s):
    """
    Precondition : string s
    Postcondition : prints the letters in s in decreasing order of frequency
    
    >>> most_frequent('aaaabbbccd')
    a
    b
    c
    d
    
    >>> most_frequent('greenteapress')
    e
    s
    r
    t
    p
    n
    g
    a
    
    """
    s_histogram = histogram(s)
    new_s = []
    for key,value in s_histogram.items():
        new_s.append((value,key))
    new_s.sort()
    new_s.reverse()
    for key,value in new_s:
        print (value)
        

def histogram(s): # from a previous excercise
    """
    Precondition : string s
    Postcondition : Returns a dictionary of characters and frequencies.
    
    >>> histogram('aaaaaaaa')
    {'a': 8}
    
    >>> {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1} == histogram('abcdef')
    True
    
    >>> pprint.pprint(histogram('abcdef'))  #pprint sorts the dictionary for us
    {'a': 1, 'b': 1, 'c': 1, 'd': 1, 'e': 1, 'f': 1}
    """
    d = dict()
    for c in s:
        d[c] = d.get(c,0)+1
    return d


doctest.run_docstring_examples(most_frequent, globals(),verbose=True)

Finding tests in NoName
Trying:
    most_frequent('aaaabbbccd')
Expecting:
    a
    b
    c
    d
ok
Trying:
    most_frequent('greenteapress')
Expecting:
    e
    s
    r
    t
    p
    n
    g
    a
ok


### Challenge: Exercise 12.4   (optional)

From a [Car Talk Puzzler](http://www.cartalk.com/content/puzzlers):

What is the longest English word, that remains a valid English word, as you remove its letters one at a time?

Now, letters can be removed from either end, or the middle, but you can’t rearrange any of the letters. Every time you drop a letter, you wind up with another English word. If you do that, you’re eventually going to wind up with one letter and that too is going to be an English word—one that’s found in the dictionary. I want to know what’s the longest word and how many letters does it have?

I’m going to give you a little modest example: Sprite. Ok? You start off with sprite, you take a letter off, one from the interior of the word, take the r away, and we’re left with the word spite, then we take the e off the end, we’re left with spit, we take the s off, we’re left with pit, it, and I. 

Write a program to find all words that can be reduced in this way, and then find the longest one.

This exercise is a little more challenging than most, so here are some suggestions:

- You might want to write a function that takes a word and computes a list of all the words that can be formed by removing one letter. These are the “children” of the word.
- Recursively, a word is reducible if any of its children are reducible. As a base case, you can consider the empty string reducible.
- The word list from [Chapter 9.1](http://www.greenteapress.com/thinkpython2/html/thinkpython2010.html) Exercise 1 doesn’t contain single letter words. So you might want to add “I”, “a”, and the empty string.
- To improve the performance of your program, you might want to memoize the words that are known to be reducible.

Allen's solution: http://greenteapress.com/thinkpython2/code/reducible.py.

In [1]:
import doctest


def get_children(word):
    """
    precondition : a string word
    postcondition : Returns all children of word

    >>> get_children('sprite')
    ['spite', 'sprit']

    >>> get_children('spite')
    ['site', 'spit']
    """
    children = []
    for i in range(len(word)):
        if is_this_word(word[0:i]+word[i+1:]):
            children.append(word[0:i]+word[i+1:])
    return children


def is_reducible(word):
    """
    precondition : string word
    postcondition : returns True if children are reducible

    >>> is_reducible('sprite')
    True

    >>> is_reducible('asdf')
    False
    """
    if word in ['', 'a', 'i']:  # base case
        return True
    if word in memo:  # checks if the word is in memo
        return memo[word]
    children = get_children(word)
    for child in children:
        memo[child] = is_reducible(child)
        return memo[child]  # memoization
    return False


def is_this_word(word):
    """
    preconditon : a string word
    postcondition : returns True if word is an actual word in word.txt
                    returns False if not

    >>> is_this_word('asdfdsafdasff')
    False

    >>> is_this_word('sprite')
    True
    """
    if word in words_dict:
        return True
    else:
        return False


def init_english_dict(file_name):
    """
    precondition : the directory of the words.txt file_name
    postcondition : returns a dictionary of words in words.txt + 'a','i',' ''
    """
    english_words = dict()
    f = open(file_name)
    lines = f.readlines()
    for l in lines:
        word = l.strip().lower()
        english_words[word] = None
    f.close()
    for word in ['a', 'i', '']:
        english_words[word] = None
    return english_words


if __name__ == "__main__":
    words_dict = init_english_dict("/home/slyu/Documents/softdes/words.txt")  #this won't work on other computers
    memo = dict()
    reducible_word = []
    for word in words_dict:
        if is_reducible(word):
            reducible_word.append(word)
    reducible_word.sort(key=len, reverse=True) #reducible_word contains all the reducible words
    print("longest reducible word : " + reducible_word[0])
    doctest.testmod()
 

longest reducible word : complecting


## Reading Journal feedback

[Please complete this short survey](https://docs.google.com/forms/d/e/1FAIpQLScQekhUrf6YYjpfQiAAbavLIA-IJklv_PX1BWbGgxj7JPolmw/viewform?c=0&w=1)

If you have any comments on this Reading Journal, feel free to leave them in the survey linked above. This could include suggestions to improve the exercises, topics you'd like to see covered in class next time, or other feedback.

If you have Python questions or run into problems while completing the reading, you should post them to Piazza instead so you can get a quick response before your journal is submitted.